### ISSUE TITLE
more food for thought

### ISSUE THREAD

**Issue by smuemd** [foxdonut/meiosis#32]
https://github.com/foxdonut/meiosis/blob/d5c8b95d4247c2d79b90e1741c612ecaaf68d56a/helpers/routing/state/index.js#L47-L62

Dear @foxdonut

I have been taking a look at your routing WIP. Array based, programmable routes are 🔥. Here are some thoughts that may make the concept easier to grok:

Initially I had problems wrapping my head around what `route` and `routes` mean or refer to in different places of the code.

We have `route` in `state`: This can be thought of an array of `...routeSegments` that match against components in the view tree and to some extend against url paths as well.

Then there is the `route` component prop (object), that gets calculated based on a given `route` array in `state`: This is where the actual routing or navigation is happening on a per component basis. Confusingly, it contains a `routes` value which references the original `route` array.

Then on top of this there are various helper functions where it is hard to tell if they work with the component prop `route` or the state `route` / `routes` or both. (e. g. `initRoute` and `nextRoute` produce a prop, while `parentRoute`, `childRoute`, `siblingRoute` produce a state route arry.

I think the confusion may be resolved by re-naming the state prop to something different, like `routing` or `navigation` or something similar. 

I would also suggest to better differentiate the two functions `nextRoute` and `initRoute` that each yield the routing prop from the other functions that yield state route arrays.

Consider the following:

```js
export function Routing (route = [], index = 0) {
  return {
    route,
    index,
    local: route[index] || {},
    child: route[index + 1] || {},
    next () {
      return Routing(route, index + 1)
    }
  }
}
```

And then in the `Root`

```jsx
import { Route, Routing } from "somewhere"
import { Beer, Beverages } from 'somewhere-else'

const componentMap = { Beer, Beverages }

function Root ({ state, actions }) {
  const routing = Routing(state.route.current)
  const Component = componentMap[routing.local.id]

  return <div>
     <Button onClick={() => actions.navigateTo([ Route.Beer(), Route.Beverages() ])}>
        Show list of beers
     <Button>
     <Component state={state} actions={actions} routing={routing} />
  </div>
}
```

And then in `Beer`

```jsx
import { Beverages } from "../beverages";
import { Beverage } from "../beverage";

const componentMap = {
  Beverages,
  Beverage
}
function Beer ({ state, actions, routing }) {
  const Component = componentMap[routing.child.id]

  return (
    <div>
      <div>Beer Page</div>
      <Component
        state={state}
        actions={actions}
        routing={routing.next()}
        beverages="beers" />
    </div>
  ) 
}
```

I think the mechanism will be much easier to understand just by virtue of better differentiating its moving parts.

this might also help clarify the signatures of the other helper functions that consume the routing prop to yield a new route array. (So that you don't end up referencing a `routes` when in fact there is only `route`.

```js
const routing = new Routing([ Route.Beer(), Route.Beverages() ])

function showBeerDetails () {
  actions.navigateTo(siblingRoute(routing, [ Route.Beverage({ id: 'b2' }) ])
}

/**
 *
 * @param {Object} routing - the component routing object / prop
 * @param {Array.<{ id: String, params: Object}>} routing.route - original state route ref
 * @param {Number} routing.index - current route segment reference
 *
 * @param {{ id: String, params: Object}} sibling - route segment to inject into route
 * @param {Array.<{ id: String, params: Object }>} [children] - optional sibling children
 *
 * @return {Array.<{ id: String, params: Object }>} - new state route
 */
function siblingRoute (routing = {}, [ sibling, ...children ]) {
  return routing.route.slice(0, routing.index).concat(sibling, children)
}
```

A am about to try your new routing mechanism with an app of mine where I have to deal with ledgers, that contain accounts, and accounts that contain transactions, and transactions that have lots of transaction details and reference different accounts each. So this will be awesome!

**Comment by foxdonut** [comment]
Once again @smuemd your feedback is very, very helpful and I appreciate it greatly.

It's great to hear that you like the array-based routing, programmable routes, and relative navigation. It's been something I've worked on for a long time, with a handful of unsatisfactory solutions before I finally arrived at this which I think has potential.

Indeed the names needed work (Naming Is Hard) but you have saved me some effort with your suggestions. I really like them and will incorporate them into the code.

As you may have already seen, I am also working on moving the utility/reusable parts of the routing code into helper functions.

For actual path/router logic,My goal is that all you need in your application is to define your routing config and to plug in the router of your choice. I'm experimenting with `feather-route-matcher` as a proof-of-concept that you really only need a simple router. I still plan to also provide some support for Mithril's router since support for it is often requested.

Thank you and keep the feedback and suggestions coming, they are :1st_place_medal:

**Comment by smuemd** [comment]
> It's been something I've worked on for a long time, with a handful of unsatisfactory solutions before I finally arrived at this which I think has potential.

Yes, I've been silently following your commits into the `docs/code` section for a while and aware of the various iterations. Routing in SPA in real hard. But this is coming together in a really nice way.

> My goal is that all you need in your application is to define your routing config and to plug in the router of your choice. I'm experimenting with feather-route-matcher as a proof-of-concept that you really only need a simple router.

Actually, I still have some meiosis routing code for url-mapper lying around. I will rework this to into another drop-in example based on your new routing helpers and send it over. (probably next weekend or so). `url-mapper` is nice for someone who needs extra special powers for query strings. The feather matcher is 5 x lighter in code size though, so probably a better first choice example.

**Comment by foxdonut** [comment]
@smuemd as per your suggestions I now have (in `routing-common-core` and only the `navigation` example for now):

- `Routing` for the component prop
- a *route* is an array of *route segments*
- `findRoute` changed to `findRouteSegments`
- `createRoute` changed to `createRouteSegments`. In the example though I still assigned the result to `Route` because `[Route.A(), Route.B()]` is still nicer than `[RouteSegment.A(), RouteSegment.B()]`. Hopefully that slight inconsistency is acceptable.
- Besides `next()`, I also added `parentRoute`, `childRoute`, and `siblingRoute` to `Routing`. Less things to import.
- Also in `Routing`, I changed `local` and `child` to `localSegment` and `childSegment` for clarity.

I'm glad you mentioned `url-mapper`, I was also thinking of using it as another "plug in your router" example.
